tokio/src/sync/oneshot.rs

```diff
@@ -1031,11 +1031,38 @@ impl State {
     }
 
     fn set_complete(cell: &AtomicUsize) -> State {
-        // TODO: This could be `Release`, followed by an `Acquire` fence *if*
-        // the `RX_TASK_SET` flag is set. However, `loom` does not support
-        // fences yet.
-        let val = cell.fetch_or(VALUE_SENT, AcqRel);
-        State(val)
+        // This method is a compare-and-swap loop rather than a fetch-or like
+        // other `set_$WHATEVER` methods on `State`. This is because we must
+        // check if the state has been closed before setting the `VALUE_SENT`
+        // bit.
+        //
+        // We don't want to set both the `VALUE_SENT` bit if the `CLOSED`
+        // bit is already set, because `VALUE_SENT` will tell the receiver that
+        // it's okay to access the inner `UnsafeCell`. Immediately after calling
+        // `set_complete`, if the channel was closed, the sender will _also_
+        // access the `UnsafeCell` to take the value back out, so if a
+        // `poll_recv` or `try_recv` call is occurring concurrently, both
+        // threads may try to access the `UnsafeCell` if we were to set the
+        // `VALUE_SENT` bit on a closed channel.
+        let mut state = cell.load(Ordering::Relaxed);
```

Comment:
This seems safer.
```suggestion
        let mut state = cell.load(Ordering::Acquire);
```

Reply:
I don't think it's _necessary_; the `compare_exchange` has an `Acquire` failure ordering, so when we actually go to set the value, we will always perform an `Acquire` operation, and if the initial `Relaxed` load is stale, the CAS will fail, and we'll loop again with a value that was accessed with an `Acquire` ordering. So, making the initial load `Relaxed` really just serves to avoid an unnecessary `Acquire` in the case that we're the only thread writing to the state; if it's contended and the `Relaxed` load is stale, we will always perform an `Acquire` on subsequent iterations.

Other CAS loops elsewhere in Tokio follow a similar pattern: https://github.com/tokio-rs/tokio/blob/03969cdae7674681d1b10926e6a56fbb8908dbb8/tokio/src/time/driver/entry.rs#L167-L192

However, I'm happy to change this if you prefer.